File: python_backend/src/utils/pagination.py
"""
Pagination utilities.
"""
from typing import Dict, Any


def get_pagination(filters: dict) -> dict:
    """
    Extract pagination parameters from filters.
    
    Args:
        filters: Dictionary with page, limit/per_page parameters
        
    Returns:
        Dictionary with page, per_page, and offset
    """
    page = int(filters.get('page', 1))
    per_page = int(filters.get('limit', filters.get('per_page', 10)))
    
    # Ensure valid values
    page = max(1, page)
    per_page = max(1, min(100, per_page))
    
    offset = (page - 1) * per_page
    
    return {
        'page': page,
        'per_page': per_page,
        'offset': offset
    }


def get_pagination_meta(total: int, page: int, per_page: int) -> Dict[str, Any]:
    """
    Generate pagination metadata.
    
    Args:
        total: Total number of records
        page: Current page number
        per_page: Records per page
        
    Returns:
        Dictionary with pagination metadata
    """
    total_pages = (total + per_page - 1) // per_page
    
    return {
        'page': page,
        'per_page': per_page,
        'total': total,
        'total_pages': total_pages,
        'has_next': page < total_pages,
        'has_prev': page > 1
    }

File: archive/backend_typescript/src/utils/pagination.ts
import { PaginationParams, PaginationMeta } from '../types';

export const getPagination = (params: PaginationParams) => {
  const page = Math.max(1, params.page || 1);
  const perPage = Math.min(100, Math.max(1, params.perPage || 10));
  const offset = (page - 1) * perPage;

  return { page, perPage, offset };
};

export const getPaginationMeta = (
  total: number,
  page: number,
  perPage: number
): PaginationMeta => {
  return {
    page,
    perPage,
    total,
    totalPages: Math.ceil(total / perPage)
  };
};
